test/e2e/upgrades/daemonsets.go

```diff
@@ -0,0 +1,197 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"k8s.io/kubernetes/pkg/api/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	extensions "k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+// DaemonSetUpgradeTest tests that a DaemonSet is running before, during and after
+// a cluster upgrade.
+type DaemonSetUpgradeTest struct {
+	daemonSet *extensions.DaemonSet
+}
+
+// Setup creates a DaemonSet and verifies that it's running
+func (t *DaemonSetUpgradeTest) Setup(f *framework.Framework) {
+	namespace := "daemonset-upgrade"
+	daemonSetName := "ds1"
+	labelSet := map[string]string{"ds-name": daemonSetName}
+	image := "gcr.io/google_containers/serve_hostname:v1.4"
+
+	// Grab a unique namespace so we don't collide.
+	ns, err := f.CreateNamespace(namespace, nil)
+	framework.ExpectNoError(err)
+
+	t.daemonSet = &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns.Name,
+			Name:      daemonSetName,
+		},
+		Spec: extensions.DaemonSetSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labelSet,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  daemonSetName,
+							Image: image,
+							Ports: []v1.ContainerPort{{ContainerPort: 9376}},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	By("Creating a DaemonSet")
+	if t.daemonSet, err = f.ClientSet.Extensions().DaemonSets(ns.Name).Create(t.daemonSet); err != nil {
+		framework.Failf("unable to create test DaemonSet %s: %v", t.daemonSet.Name, err)
+	}
+
+	By("Waiting for DaemonSet pods to become ready")
+	wait.Poll(framework.Poll, framework.PodStartTimeout, func() (bool, error) {
+		res, err := checkRunningOnAllNodes(f, t.daemonSet.Namespace, t.daemonSet.Labels)
+		framework.ExpectNoError(err)
+		return res, err
+	})
+
+	By("Validating the DaemonSet after creation")
+	t.validateRunningDaemonSet(f)
+}
+
+// Test validates that the DaemonSet is running during the upgrade (if applicable) and
+// verifies again post-upgrade that it's running
+func (t *DaemonSetUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	testDuringDisruption := upgrade == MasterUpgrade
+
+	if testDuringDisruption {
+		By("validating the DaemonSet is still running during upgrade")
+
+		wait.Until(func() {
+			t.validateRunningDaemonSet(f)
```

Comment:
I'm not sure you can do this during a disruption. The master will be down at some point and your API calls will fail.

Other upgrade tests only do testing during disruption that don't involve calling the API. Like the service test  that accesses the service through a load balancer.